docs/architecture/adr-070-unordered-account.md

```diff
@@ -2,214 +2,137 @@
-  is used, CometBFT will reject the duplicate for you.
-* For applications that leverage ABCI++, `ProcessProposal` should evaluate and reject
-  malicious proposals with duplicate transactions.
-* For applications that leverage their own application mempool, their mempool should
-  reject the duplicate for you.
-* Finally, worst case if the duplicate transaction is somehow selected for a block
-  proposal, 2nd and all further attempts to evaluate it, will fail during `DeliverTx`,
-  so worst case you just end up filling up block space with a duplicate transaction.
-
-```golang
-type TxHash [32]byte
+We facilitate replay protection by storing the unordered sequence, in the Cosmos SDK KV store. Upon transaction ingress, we check if the transaction's unordered
+sequence exists in state, or if the TTL value is stale, i.e. before the current block time. If so, we reject it. Otherwise,
+we add the unordered sequence to state. This section of the state will belong to the `x/auth` module.
 
-const PurgeLoopSleepMS = 500
+The state is evaluated during x/auth's `PreBlocker`. All transactions with an unordered sequence earlier than the current block time
+will be deleted.
 
-// UnorderedTxManager contains the tx hash dictionary for duplicates checking,
-// and expire them when block production progresses.
-type UnorderedTxManager struct {
-  // blockCh defines a channel to receive newly committed block time
-  blockCh chan time.Time
-
-  mu sync.RWMutex
-	// txHashes defines a map from tx hash -> TTL value, which is used for duplicate
-	// checking and replay protection, as well as purging the map when the TTL is
-	// expired.
-	txHashes map[TxHash]time.Time
-}
-
-func NewUnorderedTxManager() *UnorderedTxManager {
-  m := &UnorderedTxManager{
-		blockCh:  make(chan time.Time, 16),
-		txHashes: make(map[TxHash]time.Time),
-  }
-
- return m
-}
-
-func (m *UnorderedTxManager) Start() {
-  go m.purgeLoop()
-}
-
-func (m *UnorderedTxManager) Close() error {
-  close(m.blockCh)
-  m.blockCh = nil
-  return nil
+```go
+func (am AppModule) PreBlock(ctx context.Context) (appmodule.ResponsePreBlock, error) {
+	err := am.accountKeeper.GetUnorderedTxManager().RemoveExpired(sdk.UnwrapSDKContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	return &sdk.ResponsePreBlock{ConsensusParamsChanged: false}, nil
 }
+```
 
-func (m *UnorderedTxManager) Contains(hash TxHash)  bool{
-  m.mu.RLock()
-  defer m.mu.RUnlock()
-
-  _, ok := m.txHashes[hash]
-  return ok
-}
+```golang
+package keeper
 
-func (m *UnorderedTxManager) Size() int {
-  m.mu.RLock()
-  defer m.mu.RUnlock()
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
 
-  return len(m.txHashes)
-}
+	"cosmossdk.io/collections"
+	"cosmossdk.io/core/store"
+)
 
-func (m *UnorderedTxManager) Add(hash TxHash, expire time.Time) {
-  m.mu.Lock()
-  defer m.mu.Unlock()
+var (
+	// just arbitrarily picking some upper bound number.
+	unorderedSequencePrefix = collections.NewPrefix(90)
```

Comment:
what does this mean

Reply:
theres existing prefixes in the store, and i'm just picking a higher number to stay out of their way. maybe if unordered grows, it can occupy all 9x prefixes. 🤷 but yeah, mostly arbitrary